lib/ecto/adapters/sql.ex

```diff
@@ -596,6 +602,23 @@ defmodule Ecto.Adapters.SQL do
     end
   end
 
+  ## Migrations
+
+  @doc false
+  def lock_for_migrations(repo, opts, fun) do
+    alias Ecto.Migration.SchemaMigration
+    import Ecto.Query, only: [from: 2]
+
+    transaction(repo, opts, fn ->
+      from(p in {SchemaMigration.get_source(repo), SchemaMigration}, select: p.version)
+      |> Map.put(:prefix, opts[:prefix])
+      |> Map.put(:lock, "FOR UPDATE")
+      |> repo.all(timeout: :infinity, log: false)
+
+      fun.()
```

Comment:
I was thinking we could call this `fun.()` with the versions as argument. What do you think?

Reply:
No problem with that. It'll reduce some of the duplication in the functions passed to this.